typing/predef.ml

```diff
@@ -251,7 +258,7 @@ let build_initial_env add_type add_extension empty_env =
   |> add_extension ident_sys_error [type_string]
   |> add_extension ident_undefined_recursive_module
        [newgenty (Ttuple[type_string; type_int; type_int])]
-  |> add_extension ident_unhandled []
+  |> add_extension ident_unhandled_effect [type_eff (newgenty (Tvar None))]
```

Comment:
This seems to be incorrect. `add_extension` will set "`res`" in [constructor_declaration](https://v2.ocaml.org/releases/4.14/api/compilerlibref/Parsetree.html#TYPEconstructor_declaration) to `None`, but I believe it should to be essentially `Some "exn"`. In my (incomplete) attempt to implement `Unhandled_effect`, I added an optional argument to `add_extension` to make it work that way.

Reply:
I'm not an expert on this. All other pre-defined exceptions use `add_extension` as is. Why should the `Unhandled_effect` exception be different?

Reply:
Because all other exceptions are using the "standard" constructor definition
```
type exn += Constructor of argument_type
```
but this one, I believe, is (or should be) using
```
type exn += Constructor : 'a . argument_type -> exn
```
`Unhandled_effect` is unique because it will be the only built-in exception that is polymorphic.

Reply:
It may be easiest if you are able to suggest the patch for this one. Perhaps someone who understands this code can confirm this change. @trefis @lpw25?

Reply:
To be precise, I was talking about this line in `predef.ml`:
```ocaml
ext_ret_type = None;
```
It should probably be
```ocaml
ext_ret_type = if use_gadt then Some type_exn else None;
```
and then the `add_extension` call for `Unhandled_effect` should be `add_extension ~use_gadt:true`. `gadt` here stands for [generalized ADTs](https://v2.ocaml.org/manual/gadts-tutorial.html).

Reply:
GitHub does not allow me to make "suggestions" far away from the diff ranges :sweat_smile:

Reply:
I can confirm that this a soundness issue to not have the return type for `Unhandled_effect`. Without this return type, the constructor is considered as not-generalized which means that the following unsafe cast is not rejected by the typechecker
```ocaml
let cast: type a b. a eff -> b eff  = fun e ->
  try raise (Unhandled_effect e) with Unhandled_effect e -> e
```

Reply:
I made a "PR to a PR" to work around the GitHub limitation. This is my suggested changes: https://github.com/kayceesrk/ocaml/pull/9

Reply:
Merged. Thanks.